File: pkg/dev_compiler/test/codegen/closure.dart
// compile options: --closure-experimental --destructure-named-params --modules=es6
library test;
import 'dart:js';

List/*<T>*/ generic_function/*<T>*/(List/*<T>*/ items, dynamic/*=T*/ seed) {
  var strings = items.map((i) => "$i").toList();
  return items;
}

typedef void Callback({int i});

class Foo<T> {
  final int i;
  bool b;
  String s;
  T v;

  Foo(this.i, this.v);

  factory Foo.build() => new Foo(1, null);

  untyped_method(a, b) {}

  T pass(T t) => t;

  String typed_method(
      Foo foo, List list,
      int i, num n, double d, bool b, String s,
      JsArray a, JsObject o, JsFunction f) {
    return '';
  }

  optional_params(a, [b, int c]) {}

  static named_params(a, {b, int c}) {}

  nullary_method() {}

  function_params(int f(x, [y]), g(x, {String y, z}), Callback cb) {
    cb(i: i);
  }

  run(List a, String b, List c(String d), List<int> e(f(g)), {Map<Map, Map> h}) {}

  String get prop => null;
  set prop(String value) {}

  static String get staticProp => null;
  static set staticProp(String value) {}

  static const String some_static_constant = "abc";
  static final String some_static_final = "abc";
  static String some_static_var = "abc";
}

class Bar {}

class Baz extends Foo<int> with Bar {
  Baz(int i) : super(i, 123);
}

void main(args) {}

const String some_top_level_constant = "abc";
final String some_top_level_final = "abc";
String some_top_level_var = "abc";

File: pkg/dev_compiler/test/codegen/expect/closure.js
export const closure = Object.create(null);
import { core, js, dart, dartx } from 'dart_sdk';
closure.generic_function = function(T) {
  return (items: core.List<T>, seed: T): core.List<T> => {
    let strings = items[dartx.map](core.String)(dart.fn((i: T): string => `${i}`, core.String, [T]))[dartx.toList]();
    return items;
  };
};
dart.fn(closure.generic_function, T => [core.List$(T), [core.List$(T), T]]);
closure.Callback = dart.typedef('Callback', () => dart.functionType(dart.void, [], {i: core.int}));
closure.Foo$ = dart.generic(T => {
  class Foo<T> extends core.Object {
    i: number;
    b: boolean;
    s: string;
    v: T;
    static some_static_constant: string;
    static some_static_final: string;
    static some_static_var: string;
    Foo(i: number, v: T) {
      this.i = i;
      this.v = v;
      this.b = null;
      this.s = null;
    }
    static build() {
      return new (closure.Foo$(T))(1, null);
    }
    untyped_method(a, b) {}
    pass(t: T) {
      dart.as(t, T);
      return t;
    }
    typed_method(foo: closure.Foo<any>, list: core.List<any>, i: number, n: number, d: number, b: boolean, s: string, a: any[], o: Object, f: Function) {
      return '';
    }
    optional_params(a, b = null, c: number = null) {}
    static named_params(a, {b = null, c = null}: {b?: any, c?: number} = {}) {}
    nullary_method() {}
    function_params(f: (x: any, y?: any) => number, g: (x: any, opts?: {y?: string, z?: any}) => any, cb: closure.Callback) {
      dart.as(f, dart.functionType(core.int, [dart.dynamic], [dart.dynamic]));
      dart.as(g, dart.functionType(dart.dynamic, [dart.dynamic], {y: core.String, z: dart.dynamic}));
      cb({i: this.i});
    }
    run(a: core.List<any>, b: string, c: (d: string) => core.List<any>, e: (f: (g: any) => any) => core.List<number>, {h = null}: {h?: core.Map<core.Map<any, any>, core.Map<any, any>>} = {}) {
      dart.as(c, dart.functionType(core.List, [core.String]));
      dart.as(e, dart.functionType(core.List$(core.int), [dart.functionType(dart.dynamic, [dart.dynamic])]));
    }
    get prop() {
      return null;
    }
    set prop(value: string) {}
    static get staticProp() {
      return null;
    }
    static set staticProp(value: string) {}
  }
  dart.setSignature(Foo, {
    constructors: () => ({
      Foo: [closure.Foo$(T), [core.int, T]],
      build: [closure.Foo$(T), []]
    }),
    methods: () => ({
      untyped_method: [dart.dynamic, [dart.dynamic, dart.dynamic]],
      pass: [T, [T]],
      typed_method: [core.String, [closure.Foo, core.List, core.int, core.num, core.double, core.bool, core.String, js.JsArray, js.JsObject, js.JsFunction]],
      optional_params: [dart.dynamic, [dart.dynamic], [dart.dynamic, core.int]],
      nullary_method: [dart.dynamic, []],
      function_params: [dart.dynamic, [dart.functionType(core.int, [dart.dynamic], [dart.dynamic]), dart.functionType(dart.dynamic, [dart.dynamic], {y: core.String, z: dart.dynamic}), closure.Callback]],
      run: [dart.dynamic, [core.List, core.String, dart.functionType(core.List, [core.String]), dart.functionType(core.List$(core.int), [dart.functionType(dart.dynamic, [dart.dynamic])])], {h: core.Map$(core.Map, core.Map)}]
    }),
    statics: () => ({named_params: [dart.dynamic, [dart.dynamic], {b: dart.dynamic, c: core.int}]}),
    names: ['named_params']
  });
  return Foo;
});
closure.Foo = closure.Foo$();
/** @final {string} */
closure.Foo.some_static_constant = "abc";
/** @final {string} */
closure.Foo.some_static_final = "abc";
/** @type {string} */
closure.Foo.some_static_var = "abc";
closure.Bar = class Bar extends core.Object {};
closure.Baz = class Baz extends dart.mixin(closure.Foo$(core.int), closure.Bar) {
  Baz(i: number) {
    super.Foo(i, 123);
  }
};
dart.setSignature(closure.Baz, {
  constructors: () => ({Baz: [closure.Baz, [core.int]]})
});
closure.main = function(args): void {
};
dart.fn(closure.main, dart.void, [dart.dynamic]);
/** @final {string} */
closure.some_top_level_constant = "abc";
/** @final {string} */
closure.some_top_level_final = "abc";
/** @type {string} */
closure.some_top_level_var = "abc";
